spec/mongoid/contextual/mongo/documents_loader_spec.rb

```diff
@@ -0,0 +1,135 @@
+# frozen_string_literal: true
+
+require "spec_helper"
+
+describe Mongoid::Contextual::Mongo::DocumentsLoader do
+  let(:view) do
+    double('view').tap do |view|
+      allow(view).to receive(:map)
+    end
+  end
+
+  let(:klass) do
+    double
+  end
+
+  let(:criteria) do
+    double('criteria').tap do |criteria|
+      allow(criteria).to receive(:inclusions).and_return([])
+    end
+  end
+
+  let(:executor) do
+    described_class.executor
+  end
+
+  let(:subject) do
+    described_class.new(view, klass, criteria, executor: executor)
+  end
+
+  context 'state management' do
+    let(:executor) do
+      # Such executor will never execute a task, so it guarantees that
+      # our task will stay in its initial state.
+      Concurrent::ThreadPoolExecutor.new(
+        min_threads: 0,
+        max_threads: 0,
+      )
+    end
+
+    describe '#initialize' do
+      it 'initializes in pending state' do
+        expect(subject.pending?).to be_truthy
+        expect(subject.started?).to be_falsey
+      end
+    end
+
+    describe '#unschedule' do
+      it 'changes state' do
+        subject.unschedule
+        expect(subject.pending?).to be_falsey
+        expect(subject.started?).to be_falsey
+      end
+    end
+
+    describe '#execute' do
+      it 'does not change state' do
+        prev_started = subject.started?
+        prev_pending = subject.pending?
+        subject.execute
+        expect(subject.started?).to eq(prev_started)
+        expect(subject.pending?).to eq(prev_pending)
+      end
+    end
+
+    context 'when the task is completed' do
+      let(:executor) do
+        Concurrent::ImmediateExecutor.new
+      end
+
+      it 'changes the state to started' do
+        subject.wait!
+        expect(subject.started?).to be_truthy
```

Comment:
Explain this to me, the state is started even though it's finished? Is there a flag that tells you the execution is finished, or do we not need that...

Reply:
There is no `finished` state. It can be `pending`, `started`, and `cancelled`. 

This whole `DocumentLoader` class is basically a wrapper around [Concurrent::Promises::Future](https://ruby-concurrency.github.io/concurrent-ruby/1.1.4/Concurrent/Promises/Future.html). On top of it we need to know whether execution of the future has been already started or not. Hence additional state.

If the execution has been started, we just use `value!` method to get the results, or wait until the results are available. If the execution has not been started, we "cancel" the future.